### ISSUE TITLE
Enum variants are suggested in RHS of match arm

### ISSUE THREAD

**Issue by jonas-schievink** [rust-lang/rust-analyzer#12264]
```rust
enum MyEnum {
    VariantA,
    VariantB,
}

fn f(e: MyEnum) -> Option<u8> {
    match e {
        MyEnum::VariantA => Some(0),
        MyEnum::VariantB => $0
    }
}
```

We suggest both `MyEnum::VariantA` and `MyEnum::VariantB` here, but they don't really make any sense, since the expected type is `Option<u8>`.

**Comment by jonas-schievink** [comment]
Huh, apparently we don't complete `VariantB` here at all, I thought that was the misfiring completion:

```rust
enum MyEnum {
    VariantA,
    VariantB,
}

fn f(e: MyEnum) -> Option<u8> {
    match e {
        MyEnum::VariantA => Some(0),
        Varia$0
    }
}
```

**Comment by jonas-schievink** [comment]
Yeah the original issue was just that this code was triggering because for some reason `ctx.expected_type` is `MyEnum` there instead of the correct `Option<u8>`:

https://github.com/rust-lang/rust-analyzer/blob/b6b17c7a59538de37e34878c2c148e2ac7deeb8d/crates/ide-completion/src/completions/expr.rs#L155-L160
